src/main/kotlin/com/example/petshopkotlin/pet/PetService.kt

```
@@ -0,0 +1,40 @@
+package com.example.petshopkotlin.pet
+
+import com.example.petshopkotlin.pet.model.Pet
+import org.bson.types.ObjectId
+import org.springframework.stereotype.Service
+
+@Service
+class PetService(val repo: PetRepository) {
+    fun getPets() = repo.findAllByOrderByIdDesc()
+
+    fun addPet(pet: Pet): Pet {
+        val errors = validatePet(pet)
+        require(errors.isBlank()) { errors }
+
+        return repo.save(pet)
+    }
+
+    fun adoptPet(id: ObjectId): String {
+
+        val pet = repo.findById(id).orElseThrow { IllegalArgumentException("Pet with ID: $id doesn't exist") }
```

MongoRepository is returning a Java Optional? You can assume the value inside a Java optional is never null, so here you can immediately use Pet without question mark as a type. (Fail fast- > fail as close to the problem as possible.

---

You can also look if there is a (reliable, often used) kotlin specific implementation of MongoRepository available, that would not return java Optional but Kotlin nullable

---

Mongo repository is returning a Java Optional. Thus I use` orElseThrow ` method to throw an error if the value doesn't exist. If value exists then it will return Pet.

How can I implement it better?